packages/components-library/src/components/ui/Pagination.vue

```diff
@@ -0,0 +1,287 @@
+
+      const midPage = Math.floor(this.midPageSize / 2)
+      const isMidRange = this.val.page > midPage && !(this.pageCount < this.relativePageSize)
+      if (isMidRange) {
+        // Generate first part of the middle range.
+        for (let i = this.val.page - midPage; i < this.val.page; i++) {
+          if (i > 0) { pages.push(i) }
+        }
+        // Generate the second part of the middle range.
+        for (let i = this.val.page; i <= this.val.page + midPage; i++) {
+          if (i > 0 && i <= this.pageCount) { pages.push(i) }
+        }
+      } else {
+        // Render until pageCount or relativePageSize max.
+        const _r = (this.pageCount > this.midPageSize) ? this.midPageSize : this.pageCount
+        for (let i = 1; i <= _r; i++) pages.push(i)
+      }
+
+      // Pagination is structured like: [start] [midrange] [end]
+      // This add the start in the following situation: 1 2 3 [4] 5 6
+      // Otherwise, it would become: 2 3 [4] 5 6
+      const firstPage = pages[0]
+      if (firstPage === 2) { pages.unshift(1) }
+
+      // The same with the end part:
+      const lastPage = pages[pages.length - 1]
+      if (lastPage === (this.pageCount - 1)) { pages.push(this.pageCount) }
+
+      return pages
+    }
+  },
+  data: function () {
+    return {
+      val: { size: 20, page: 1, loading: false, count: 0 }
+    }
+  },
+  methods: {
+    async fetchData (page) {
+      // Omit the Pagination method when you're using multiple
+      // Pagination components with the same data source.
+      if (!this.method) { return }
+      page = parseInt(page, 10)
+
+      this.updateValue({ loading: true, page })
+
+      const { count } = await this.method()
+      this.updateValue({ loading: false, count, page })
+
+      this.val = { ...this.value, ...{ loading: false, count, page } }
+    },
+    navigate (page) {
+      if (!this.useRouter) { return this.fetchData(page) }
+
+      if (this.$route.query.page) {
+        if (
+          parseInt(this.$route.query.page, 10) !== page ||
+          parseInt(this.$route.query.size, 10) !== this.value.size
+        ) {
+          // (!) Merge with existing query; e.g. bookmarks
+          this.$router.push({
+            path: this.$route.path,
+            query: { ...this.$route.query, page, size: this.value.size }
+          })
+        }
+      } else {
+        this.$router.replace({
+          path: this.$route.path,
+          query: { ...this.$route.query, page, size: this.value.size }
+        })
+      }
+    },
+    updateValue (value) {
+      this.val = { ...this.val, ...this.value, ...value }
+      this.$emit('input', { ...this.value, ...this.val, ...value })
+    }
+  },
+  props: {
+    /**
+     * Add css classes to pagination buttons.
+     */
+    styles: {
+      type: Object,
+      required: false,
+      default: () => ({})
+    },
+    /**
+     * The method to retrieve items with.
+     * @returns {Object} {count: Number}
+     */
+    method: {
```

Comment:
suggest using more descriptive name , ( maybe retrieveItems, or retrieveItemsFunction, or fetchItems, or getItems )

Reply:
Yes, good point. Renamed to `fetchItems`